### PR TITLE
Add support for classrooms as destinations

### LINKED ISSUE #21: S-3.1: As a user, I want to be able to select a Concordia start location and a Concordia end destination in order to get directions.
Once the starting point and destination are entered, the directions are displayed.

Details:
1. A directions button is located at the bottom right of the screen. 
Pressing it will prompt 2 inputs for locations. The top input is for the starting location while the bottom input will take the destination input.
1. Location inputs can be a building address or a classroom number. 
1. If it’s a classroom, the building acts as the destination.
1. The available locations can only be within either the Loyola campus or the SGW campus.
1. Available Start locations are only within the Loyola campus or SGW campus

Acceptance Criteria:
- [x] Given that the map is open, when I choose a start location and a destination location, then the map displays an itinerary or a list of itineraries to choose from. 
- [x] When picking an itinerary, the map with show a list of directions and draw the path on the map. #84

### LINKED ISSUE #105: Refactor the JSON of the shuttle bus schedule
Currently the JSON is in the `shuttle_data` class, we need to refactor it and move it to assets. Improves #78 3.5

### LINKED ISSUE #107: Add individual room support and search capability
Add building rooms and search capability in the search bar.

### PR DESCRIPTION
**Pull request by andrelegault**
# Description
This pull request adds classrooms as supported destinations. It does so by first importing a list of strings (`assets/destinations.json`) and then by parsing what the user inputs. Finally, what was parsed is matched to a specific LatLng (`storage/app_constants.dart`), depending on the first letter of the input (`services/location_search.dart`).

Users can search for classrooms that will (for now, at least) lead them to one of the following:
- The SGW campus
- The Loyola campus
- The Hall building
- JMSB

Future extensions of this feature would combine both outdoor and indoor navigation into one itinerary, shown both on the map and in the directions widget.

## Related issues
Fixes #105 
Fixes #107 
Fixes #21

### CODE REVIEW

File: lib/models/itinerary.dart [hunk 1]
```diff
@@ -1,54 +1,114 @@
...
+  static Map<String, Map<String, String>> getDirectionList(
+      Map<String, dynamic> rawJson) {
+    Map<String, Map<String, String>> temp = Map<String, Map<String, String>>();
+    var steps = rawJson["routes"][0]["legs"][0]["steps"].length;
+    for (int i = 0; i < steps; i++) {
+      String s1 = rawJson["routes"][0]["legs"][0]["steps"][i]
+              ["html_instructions"]
+          .replaceAll(
+              //Regex to replace certain special characters in HTML with whitespace
+              RegExp(r'(<\/?\w+\/?>?| \w+=\"\w+-\w+:\d.\d\w+\">)'),
```
Comment (MichelRahme) [inline lib/models/itinerary.dart]: Move this regex to a constant variable.

File: lib/models/itinerary.dart [hunk 2]
```diff
@@ -1,54 +1,114 @@
...
+              ' ');
+      String s2 =
```
Comment (MichelRahme) [inline lib/models/itinerary.dart]: rename

File: lib/models/itinerary.dart [hunk 3]
```diff
@@ -1,54 +1,114 @@
...
+          rawJson["routes"][0]["legs"][0]["steps"][i]["duration"]["text"];
+      String s3 =
```
Comment (MichelRahme) [inline lib/models/itinerary.dart]: rename

File: lib/models/itinerary.dart [hunk 4]
```diff
@@ -1,54 +1,114 @@
...
-              ["distance"]["text"];
-          _itinerary.addAll({
-            s1: {s2: s3}
-          });
-        }
-        safety = false;
-      }
+  /// Returns a list of Polylines obtained by parsing a json object from Google's Directions API.
+  /// similar to [getDirectionList(Map<String, dynamic>)] but returns a list of Polyline objects instead of a Map of directions.
+  static List<Polyline> getPolylinePoints(Map<String, dynamic> rawJson) {
```
Comment (WindOfRussia) [inline lib/models/itinerary.dart]: You might wanna do some inline comments here for clarity.

File: lib/providers/shuttle_data.dart [hunk 5]
```diff
@@ -1,34 +1,48 @@
...
 
       for (var i = 0; i < length; i++) {
-        if (int.parse(_json[campus][day][i]) >= (now.hour * 100) + now.minute) {
-          shuttleTime = _json[campus][day][i];
+        if (int.parse(schedule[campus][day][i]) >=
+            (time.hour * 100) + time.minute) {
+          shuttleTime = schedule[campus][day][i];
           break;
+        } else {
+          if (campus == "loyola") {
```
Comment (MichelRahme) [inline lib/providers/shuttle_data.dart]: Can you explain what is happening here?

File: lib/providers/shuttle_data.dart [hunk 6]
```diff
@@ -1,34 +1,48 @@
...
+            if (int.parse(day) < 5) {
+              shuttleTime = "0730";
+            } else {
+              shuttleTime = "0740";
+            }
+          } else {
+            shuttleTime = "0745";
+          }
```
Comment (WindOfRussia) [inline lib/providers/shuttle_data.dart]: Again some inline comments would help clarify this.

File: lib/services/location_search.dart [hunk 7]
```diff
@@ -1,58 +1,115 @@
...
-                mapData.changeCampus('loyola');
-                mapData.changeEnd(loyola);
-                mapData.changeMode("driving");
-                Navigator.pushNamed(context, '/directions');
-              },
-            );
-          },
-        ),
-      ],
+    getClassrooms();
```
Comment (MichelRahme) [inline lib/services/location_search.dart]: Just call lines 20, 21 here and get rid of the function `getClassrooms()`

File: lib/services/location_search.dart [hunk 8]
```diff
@@ -1,58 +1,115 @@
...
+          Provider.of<MapData>(context, listen: false).controllerStarting.text =
+              "Current Location";
+          Navigator.of(context).pop();
+          Provider.of<MapData>(context, listen: false)
+              .controllerDestination
+              .text = "SGW, Montreal";
+          Provider.of<MapData>(context, listen: false).controllerStarting.text =
+              "Current Location";
+          Provider.of<MapData>(context, listen: false).changeStart(
+              Provider.of<MapData>(context, listen: false).getCurrentLocation);
```
Comment (WindOfRussia) [inline lib/services/location_search.dart]: reusing a variable for this provider would be preferable.
```mapData = Provider.of ....```

File: lib/services/location_search.dart [hunk 9]
```diff
@@ -1,58 +1,115 @@
...
+          switch (suggestionList[index][0].toString()) {
```
Comment (WindOfRussia) [inline lib/services/location_search.dart]: should we extract this algorithm into its own method?

File: lib/widgets/directions_widget.dart [hunk 10]
```diff
@@ -21,20 +20,17 @@ class _DirectionsWidgetState extends State<DirectionsWidget> {
 
     ///Memoizer used to optimize the List Builder
```
Comment (MichelRahme) [inline lib/widgets/directions_widget.dart]: You don't need the `Memoizer` anymore, it's not being used.

File: lib/widgets/directions_widget.dart [hunk 11]
```diff
@@ -21,20 +20,17 @@ class _DirectionsWidgetState extends State<DirectionsWidget> {
...
+      Map<String, Map<String, String>> test =
+          Provider.of<MapData>(context, listen: false)?.itinerary?.itinerary;
+      await Future.delayed(Duration(seconds: 1));
+      return Future.value(test);
     }
 
     return Expanded(
       child: FutureBuilder(
-          future: _fetchData(),
+          future: _fetchMoreData(),
```
Comment (MichelRahme) [inline lib/widgets/directions_widget.dart]: Rename this, name `_fetchMoreData()` is vague

File: lib/widgets/shuttle_tile.dart [hunk 12]
```diff
@@ -10,46 +10,60 @@ import 'package:google_fonts/google_fonts.dart';
...
-            fontWeight: FontWeight.w600,
-            color: constants.blackColor,
-          ),
-        ),
-      ),
-    );
+    return FutureBuilder(
+        future: Provider.of<ShuttleData>(context, listen: false)
+            .getNextShuttle(campus),
+        builder: (context, AsyncSnapshot itinerary) {
```
Comment (MichelRahme) [inline lib/widgets/shuttle_tile.dart]: Not `itinerary`, change to `shuttleTime`

File: lib/widgets/shuttle_tile.dart [hunk 13]
```diff
@@ -10,46 +10,60 @@ import 'package:google_fonts/google_fonts.dart';
...
-              .getNextShuttle(campus),
-          style: GoogleFonts.raleway(
-            fontSize: 10.0,
-            fontWeight: FontWeight.w600,
-            color: constants.blackColor,
-          ),
-        ),
-      ),
-    );
+    return FutureBuilder(
```
Comment (WindOfRussia) [inline lib/widgets/shuttle_tile.dart]: Nice, cleaner.

File: lib/services/location_search.dart (hunk 9 above)
Comment (andrelegault) [reply lib/services/location_search.dart]: check https://github.com/MichelRahme/ConcordiaNavigation/blob/junk/98-directions-list/lib/providers/itinerary_data.dart i started doing that a bit there but nothing came of it

**Comment by EvanMateo7** [comment]
The failing widget test can be ignored for this PR if needed.
It may not be needed anymore as the widget it's testing has changed implementation.

image

**Comment by WindOfRussia** [comment]
@andrelegault hey, double check your fixes. looks like something in fad8ffce1b8807aaebf38d35c8d5df60a3d61979 broke the builds.

**Comment by andrelegault** [comment]
> The failing widget test can be ignored for this PR if needed.
> It may not be needed anymore as the widget it's testing has changed implementation.
> 
> image

Yeah LocationSearch widget has changed implementation

File: test/location_search_test.dart [hunk 14]
```diff
@@ -13,7 +13,7 @@ void main() {
 
       await tester.tap(find.byIcon(Icons.search));
       await tester.pumpAndSettle();
-      expect(find.text("SGW Campus, Montreal"), findsOneWidget);
+      expect(find.text("HALL BUILDING"), findsOneWidget);
```
Comment (WindOfRussia) [inline test/location_search_test.dart]: still failing. we should rewrite this test.

Comment (andrelegault) [reply test/location_search_test.dart]: I disabled the test for now, we'll have to remake the test to fit our current implementation.

Comment (EvanMateo7) [reply test/location_search_test.dart]: That's fine

**Pull request merged**